util/src/error.rs

```diff
@@ -1,79 +1,57 @@
 //! Convenient structs, implementations and types for nicer handling of our own custom error types.
 
 use core::fmt::Display;
-use std::fmt::Debug;
+use std::fmt;
+
+use failure::{Backtrace, Context, Fail};
 
 /// Generic structure for application errors that can be implemented over any error of kind `K`.
 /// Namely, `K` is intended to be a simple, C-style enum.
-///
-/// # Examples
-///
-/// ```
-/// #[derive(Debug)]
-/// enum MyOwnErrorCodes {
-///     AnError,
-///     AnotherError
-/// };
-///
-/// type MyErrorType = witnet_util::error::Error<MyOwnErrorCodes>;
-/// ```
 #[derive(Debug)]
-pub struct Error<K: Debug> {
-    kind: K,
-    message: String
+pub struct WitnetError<K: Fail> {
+    inner: Context<ErrorKind<K>>,
+}
+
+/// ErrorKind
+#[derive(Debug, Fail)]
+pub enum ErrorKind<K: Fail> {
+    /// storage error
+    #[fail(display = "StorageError: {}", 0)]
+    Storage(K),
 }
 
-/// Implement the `::new()` constructor for the generic Error<K> type.
-impl <K: Debug> Error<K> {
+impl<K: Fail> WitnetError<K> {
+    /// create storage error
+    pub fn storage_err(err: K) -> Self {
```

Comment:
The `util` crate should not know about the existence of the `storage` crate. Everything in this file is supposed to be generic.